packages/alfa-math/src/matrix.ts

```diff
@@ -1,63 +1,175 @@
-import * as math from "mathjs";
-import { Vector } from "./vector";
+import { Real } from "./real";
 
-export type Matrix<M extends number = number, N extends number = number> = [
-  Vector<N>,
-  ...Array<Vector<N>>
-] & { readonly length: M };
+/**
+ * @see https://en.wikipedia.org/wiki/Matrix_(mathematics)
+ */
+export type Matrix = Array<Array<number>>;
 
 export namespace Matrix {
-  export function of<M extends number, N extends number>(
-    matrix: Matrix<M, N>
-  ): Matrix<M, N> {
-    return matrix;
+  export function isMatrix(value: unknown): value is Matrix {
+    return (
+      Array.isArray(value) &&
+      value.every(
+        (r) => Array.isArray(r) && r.every((v) => typeof v === "number")
+      )
+    );
   }
 
-  export function identity<M extends number, N extends number>(
-    m: M,
-    n: N
-  ): Matrix<M, N> {
-    return math.identity([m, n]) as Matrix<M, N>;
+  export function isSquare(m: Matrix): boolean {
+    return rows(m) === columns(m);
   }
 
-  export function clone<M extends number, N extends number>(
-    matrix: Matrix<M, N>
-  ): Matrix<M, N> {
-    return matrix.map((vector) => Vector.clone(vector)) as Matrix<M, N>;
+  export function clone(m: Matrix): Matrix {
+    return m.map((r) => r.slice(0));
   }
 
-  export function multiply<M extends number, N extends number>(
-    matrix: Matrix<M, N>,
-    vector: Vector<N>
-  ): Vector<N>;
+  export function equals(m: Matrix, n: Matrix, e?: number): boolean {
+    return (
+      m.length === n.length &&
+      m.every(
+        (r, i) =>
+          r.length === n[i].length &&
+          r.every((v, j) => Real.equals(v, n[i][j], e))
+      )
+    );
+  }
+
+  export function identity(n: number): Matrix {
+    return new Array<Array<number>>(n).fill([]).map((_, i) => {
+      const r = new Array<number>(n).fill(0);
+      r[i] = 1;
+      return r;
+    });
+  }
+
+  export function size(m: Matrix): [number, number] {
+    return [rows(m), columns(m)];
+  }
+
+  export function rows(m: Matrix): number {
+    return m.length;
+  }
+
+  export function columns(m: Matrix): number {
+    return m.length === 0 ? 0 : m[0].length;
+  }
+
+  export function row(m: Matrix, i: number): Array<number> {
+    return m[i];
+  }
+
+  export function column(m: Matrix, i: number): Array<number> {
+    return m.map((r) => r[i]);
+  }
+
+  export function add(m: Matrix, n: Matrix): Matrix {
+    return m.map((r, i) => r.map((v, j) => v + n?.[i]?.[j]));
```

Comment:
Shouldn't that be
```suggestion
    return m.map((r, i) => r.map((v, j) => v + (n?.[i]?.[j] ?? 0)));
```
to avoid a `v + undefined` lurking around.

Reply:
That is actually very intentional as it ensures that the matrix and vector operations align with IEEE floating point behaviour. This means that:

```ts
Vector.add([1, 2, 3], [1, 2]) == [2, 4, NaN]
```